File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static boolean[] visited; 
    static int ret;
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        visited = new boolean[dungeons.length];
        go(k, 0, dungeons);
        answer = ret;
        return answer;
    }
    
    public void go(int hp, int depth, int[][] dungeons){
        if(depth > dungeons.length)
            return;
        ret = Math.max(ret, depth);
        for(int i =0; i < dungeons.length; i++){
            if(!visited[i]){
                visited[i] = true;
                if(hp >= dungeons[i][0])
                    go(hp - dungeons[i][1], depth + 1, dungeons);
                visited[i] = false;
            }
        }
        
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <vector>
#include <algorithm>

using namespace std;

int ret = 0;
bool visited[9];

void go(int depth, int k, vector<vector<int>> dungeons){
    if(k < 0)
        return;
    
    ret = max(ret, depth);
    
    for(int i = 0; i < dungeons.size(); i++){
        if(!visited[i] && k >= dungeons[i][0]){
            visited[i] = true;
            go(depth + 1, k - dungeons[i][1], dungeons);
            visited[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons){
    int answer = -1;
    go(0, k, dungeons);
    answer = ret;
    return answer;
}